io/mpi/src/TMPIFile.cxx

```diff
@@ -0,0 +1,549 @@
+   for (UInt_t f = 0; f < fClients.size(); ++f) {
+      if (fClients[f].GetFile()) {
+         R__DeleteObject(fClients[f].GetFile(), kTRUE);
+      } else {
+         // We back up the file (probably due to memory constraint)
+         TFile *file = TFile::Open(fClients[f].GetLocalName(), "UPDATE");
+         if (file->IsZombie())
+            exit(1);
+         // Remove object that can be incrementally merge and will be reset by the client code.
+         R__DeleteObject(file, kTRUE);
+         file->Write();
+         delete file;
+      }
+   }
+   fLastMerge = TTimeStamp();
+   fNClientsContact = 0;
+   fClientsContact.Clear();
+
+   return result;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// Register that a client has sent a file.
+///
+/// taken from root/tutorials/net/parallelMergeServer.C
+
+void TMPIFile::ParallelFileMerger::RegisterClient(UInt_t clientID, TFile *file)
+{
+   ++fNClientsContact;
+   fClientsContact.SetBitNumber(clientID);
+   TMPIClientInfo ntcl(std::string(fFilename).c_str(), clientID);
+   if (fClients.size() < clientID + 1) {
+      fClients.push_back(ntcl);
+   }
+   fClients[clientID].SetFile(file);
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// Return true, if enough client have reported
+///
+/// In the case of TMPIFile this happens everytime a client/worker sends
+/// the buffer (tested).
+///
+/// taken from root/tutorials/net/parallelMergeServer.C
+
+Bool_t TMPIFile::ParallelFileMerger::NeedMerge(Float_t clientThreshold)
+{
+
+   if (fClients.size() == 0) {
+      return kFALSE;
+   }
+
+   // Calculate average and rms of the time between the last 2 contacts.
+   Double_t sum = 0.;
+   Double_t sum2 = 0.;
+   for (UInt_t c = 0; c < fClients.size(); ++c) {
+      sum += fClients[c].GetTimeSincePrevContact();
+      sum2 += fClients[c].GetTimeSincePrevContact() * fClients[c].GetTimeSincePrevContact();
+   }
+   Double_t avg = sum / fClients.size();
+   Double_t sigma = sum2 ? TMath::Sqrt(sum2 / fClients.size() - avg * avg) : 0;
+   Double_t target = avg + 2 * sigma;
+   TTimeStamp now;
+   if ((now.AsDouble() - fLastMerge.AsDouble()) > target) {
+      return kTRUE;
+   }
+   Float_t cut = clientThreshold * fClients.size();
+   return fClientsContact.CountBits() > cut || fNClientsContact > 2 * cut;
+}
+
+
+
+////////////////////////////////////////////////////////////////////////////////
+/// return True if this is the Collector rank, otherwise False
+
+Bool_t TMPIFile::IsCollector()
+{
+   return !fMPILocalRank;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// Called by the Workers only: Copies the current content in memory and
+/// sends it asynchronously to the Collector for merging and writing to disk.
+///
+
+void TMPIFile::CreateBufferAndSend()
+{
+   if (this->IsCollector()) {
+      Error("CreateBufferAndSend", " should not be called by a collector");
```

Comment:
Since this probably indicate a user error, shouldn't it 'return' rather than continue the function?